### PR TITLE
test(core): add unit tests for buy-x-get-y-free promotion condition

### PR DESCRIPTION
**Pull request by GabrielRoc**
## Description

Adds a direct unit test suite for the `buy_x_get_y_free` promotion condition (`buy-x-get-y-free-condition.ts`), which previously had **0% function coverage** — the `check()` method was never invoked by any test and was only exercised transitively through integration/e2e tests.

Test cases are derived using **equivalence partitioning** and **boundary-value analysis** over the two numeric drivers in `check()`:

- `quantity = Math.floor(matches / amountX)` — the "buy X" gate
- `Math.min(line.quantity, placesToAllocate)` — the "get Y free" allocation, applied cheapest-first

12 tests covering:
- Buy-X threshold boundaries (`matches` absent / below threshold / exactly at threshold)
- Free-item candidate partitions (Y variant absent / present)
- Allocation boundaries (fits one line / spills to next / exceeds total stock)
- Cheapest-first ordering (exclude-tax prices / include-tax prices / equal prices)
- Variant overlap (variant in both X and Y sets)
- `amountY = 0` edge case (empty allocation, still truthy)

All tests are database-free, using `order-test-utils.ts` to construct `Order`/`RequestContext` in memory. The suite runs with the existing Vitest unit config (`packages/core`).

## Breaking changes

None.

## Checklist

- [x] I have set a clear title
- [x] My PR is small and contains a single feature
- [x] I have added or updated test cases

Relates to #4835

---
<a href="https://app.blacksmith.sh/vendurehq/codesmith/vendure/pr/4879"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1785127778&installation_id=128408429&pr_number=4879&repository=vendurehq%2Fvendure&return_to=https%3A%2F%2Fgithub.com%2Fvendurehq%2Fvendure%2Fpull%2F4879&signature=e20a4a8cb10eb3f67ee33afa6d663c2a5dd786efd3e093723bb90262ab12c01f"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

**Comment by GabrielRoc** [comment]
I have read the CLA Document and I hereby sign the CLA

File: packages/core/src/config/promotion/conditions/buy-x-get-y-free-condition.spec.ts [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+                amountX: 2,
+                variantIdsX: ['x1'],
+                amountY: 3,
+                variantIdsY: ['y1', 'y2'],
+            });
+
+            // Equal unit prices: the comparator returns 0, the original order is kept,
+            // so the 3 free places fill the first line then spill into the second.
+            expect(await check(ctx, order, args)).toEqual({ freeItemsPerLine: { first: 2, second: 1 } });
+        });
```
[the remark this answers is not in the document]
Comment (GabrielRoc) [reply packages/core/src/config/promotion/conditions/buy-x-get-y-free-condition.spec.ts:198]: Dismissed. `Array.prototype.sort()` has been guaranteed stable since ES2019 (V8 7.0 / Node.js 12). Vendure requires Node.js 20+, so there is no real risk of flakiness here. The test comment already documents the comparator-returns-0 semantics explicitly.

File: packages/core/src/config/promotion/conditions/buy-x-get-y-free-condition.spec.ts [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+        it('orders by unitPriceWithTax when the channel prices include tax', async () => {
+            const ctxIncl = createRequestContext({ pricesIncludeTax: true });
+            // lineA: net 100, gross 100 (no tax).  lineB: net 90, gross 135 (50% tax).
+            // By unitPrice (net):       B (90)  < A (100)  -> cheapest is B
+            // By unitPriceWithTax (gross): A (100) < B (135) -> cheapest is A
+            const lineA = line({ lineId: 'a', variantId: 'y1', quantity: 5, listPrice: 100 });
+            const lineB = line({ lineId: 'b', variantId: 'y2', quantity: 5, listPrice: 90, taxRate: 50 });
+            const order = orderWith([line({ lineId: 'x', variantId: 'x1', quantity: 2 }), lineA, lineB]);
```
[the remark this answers is not in the document]
Comment (GabrielRoc) [reply packages/core/src/config/promotion/conditions/buy-x-get-y-free-condition.spec.ts:209]: Fixed in commit `3e30e0576`. Added a comment to the fixture clarifying that `listPriceIncludesTax` is intentionally left at its default `false` so that `listPrice` represents the net price, making `unitPriceWithTax = listPrice × (1 + taxRate)`. This is what produces the ordering flip between net and gross prices that the test is designed to demonstrate — setting `listPriceIncludesTax: true` would collapse the gross prices and remove the observable contrast.

**Comment by biggamesmallworld** [comment]
Nice, thorough boundary/equivalence coverage. One gap worth closing here: there's no `amountX: 0` case — which is exactly the [OSS-580](https://github.com/vendurehq/vendure/issues/4888) division-by-zero bug (`Math.floor(matches / 0) = Infinity` bypassing the `!quantity` guard).

That bug was just fixed on `master` in #4922 (adds `if (args.amountX < 1) return false;`), so this is the natural home for the regression test:

```ts
it('returns false when amountX is 0 (no division-by-zero bypass)', async () => {
    const order = orderWith([
        line({ lineId: 'l1', variantId: 'x1', quantity: 2 }),
        line({ lineId: 'l2', variantId: 'y1', quantity: 5 }),
    ]);
    const args = buildArgs({ amountX: 0, variantIdsX: ['x1'], amountY: 1, variantIdsY: ['y1'] });
    expect(await check(ctx, order, args)).toBe(false);
});
```

**Order matters:** this branch is currently behind `master` and doesn't yet include the #4922 guard, so the assertion above will fail until you merge/rebase `master` in. Please update the branch first, then add the case — that way the test also serves as verification that the #4922 fix holds.

**Comment by biggamesmallworld** [comment]
@GabrielRoc happy to approve when the above test case is added 👍

**Comment by GabrielRoc** [comment]
Merged current master in (picks up the #4922 guard) and added the `amountX = 0` case in the `documented edge-case behaviour` block. Verified it's a real regression guard — it fails with the guard removed and passes with it in place. All 13 tests green.

**Review by biggamesmallworld — APPROVED** [review]
